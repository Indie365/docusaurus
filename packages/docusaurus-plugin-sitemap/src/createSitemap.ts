/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {ReactElement} from 'react';
import {matchRoutes, type RouteConfig as RRRouteConfig} from 'react-router-config';
import {SitemapStream, streamToPromise} from 'sitemap';
import {applyTrailingSlash} from '@docusaurus/utils-common';
import {createMatcher} from '@docusaurus/utils';
import type {DocusaurusConfig} from '@docusaurus/types';
import type {RouteConfig} from '@docusaurus/types/src/routing';
import type {HelmetServerState} from 'react-helmet-async';
import type {PluginOptions} from './options';

function isNoIndexMetaRoute({
  head,
  route,
}: {
  head: {[location: string]: HelmetServerState};
  route: string;
}) {
  const isNoIndexMetaTag = ({
    name,
    content,
  }: {
    name?: string;
    content?: string;
  }): boolean => {
    if (!name || !content) {
      return false;
    }
    return (
      // meta name is not case-sensitive
      name.toLowerCase() === 'robots' &&
      // Robots directives are not case-sensitive
      content.toLowerCase().includes('noindex')
    );
  };

  // https://github.com/staylor/react-helmet-async/pull/167
  const meta = head[route]?.meta.toComponent() as unknown as
    | ReactElement<{name?: string; content?: string}>[]
    | undefined;
  return meta?.some((tag) =>
    isNoIndexMetaTag({name: tag.props.name, content: tag.props.content}),
  );
}

export default async function createSitemap(
  siteConfig: DocusaurusConfig,
  routesPaths: string[],
  routes: RouteConfig[],
  head: {[location: string]: HelmetServerState},
  options: PluginOptions,
): Promise<string | null> {
  const {url: hostname} = siteConfig;
  if (!hostname) {
    throw new Error('URL in docusaurus.config.js cannot be empty/undefined.');
  }
  const {changefreq, priority, ignorePatterns} = options;

  const ignoreMatcher = createMatcher(ignorePatterns);

  function isRouteExcluded(route: string) {
    return (
      route.endsWith('404.html') ||
      ignoreMatcher(route) ||
      isNoIndexMetaRoute({head, route})
    );
  }

  const includedRoutes = routesPaths.filter((route) => !isRouteExcluded(route));

  if (includedRoutes.length === 0) {
    return null;
  }

  function mapRoute(route: RouteConfig) : RRRouteConfig {
    const { component, routes: originalRoutes, ...rrRoute } = route;
    return { routes: originalRoutes?.map(mapRoute), ...rrRoute };
  }

  const sitemapStream = new SitemapStream({hostname, lastmodDateOnly: options.lastmod === 'date'});
  const rrRoutes: RRRouteConfig[] = routes.map(mapRoute);

  includedRoutes.forEach((routePath) => {
    sitemapStream.write({
      url: applyTrailingSlash(routePath, {
        trailingSlash: siteConfig.trailingSlash,
        baseUrl: siteConfig.baseUrl,
      }),
      changefreq,
      priority,
      lastmod:
        routes !== undefined && options.lastmod !== false
          ? matchRoutes(rrRoutes, routePath).find(
            ({route}) => route?.exact === true,
          )?.route?.lastModified
          : undefined,
    });
  });

  sitemapStream.end();

  const generatedSitemap = (await streamToPromise(sitemapStream)).toString();

  return generatedSitemap;
}
