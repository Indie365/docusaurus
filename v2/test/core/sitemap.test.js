import '@babel/polyfill';
import createSitemap from '@lib/core/sitemap';
import loadSetup from '../loadSetup';

describe('sitemap', () => {
  test('simple site', async () => {
    const props = await loadSetup('simple');
    const sitemap = await createSitemap(props);
    expect(sitemap).toMatchSnapshot();
  });

  test('translated site', async () => {
    const props = await loadSetup('translated');
    const sitemap = await createSitemap(props);
    expect(sitemap).toMatchSnapshot();
  });

  test('versioned site', async () => {
    const props = await loadSetup('versioned');
    const sitemap = await createSitemap(props);
    expect(sitemap).toMatchSnapshot();
  });

  test('translated + versioned site', async () => {
    const props = await loadSetup('transversioned');
    const sitemap = await createSitemap(props);
    expect(sitemap).toMatchSnapshot();
  });

  test('empty site', async () => {
    const props = await loadSetup('empty');
    expect(createSitemap(props)).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Url in siteConfig.js cannot be empty/undefined"`,
    );
  });
});
