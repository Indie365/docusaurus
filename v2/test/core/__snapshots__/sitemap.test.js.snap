// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sitemap empty site 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://docusaurus.io/docs/endiliey/permalink</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url></urlset>"`;

exports[`sitemap simple site 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://docusaurus.io/docs/endiliey/permalink</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url></urlset>"`;

exports[`sitemap translated + versioned site 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://docusaurus.io/docs/en/next/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/next/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/next/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/next/endiliey/permalink</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/next/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/next/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/1.0.0/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/1.0.0/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/next/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/1.0.0/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/1.0.0/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/1.0.0/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/1.0.0/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/en/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/ko/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/en/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/ko/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url></urlset>"`;

exports[`sitemap translated site 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://docusaurus.io/docs/en/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/endiliey/permalink</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/en/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/ko/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/en/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/ko/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/en/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/ko/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url></urlset>"`;

exports[`sitemap versioned site 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://docusaurus.io/docs/next/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/next/endiliey/permalink</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/next/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/next/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/1.0.0/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/1.0.0/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/hello</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/1.0.0/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/foo/baz</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/docs/foo/bar</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/</loc><changefreq>weekly</changefreq><priority>0.5</priority></url><url><loc>https://docusaurus.io/hello/world</loc><changefreq>weekly</changefreq><priority>0.5</priority></url></urlset>"`;
